### DISCUSSION TITLE
How to obtain absolute file paths (for images)?

### DISCUSSION THREAD

**Discussion by stephanbogner** [remarkjs/remark#990] [Q&A] [answered]
### Intention:
I'd like to create a remark-plugin that converts `![Some alt](image.png)` to an `<img>`-element with a srcset using [sharp](https://www.npmjs.com/package/sharp) to convert the images.

---

### The problem:
I have no reference to the actual image file.

---

#### (A) I tried two ways:
- To access the information for the image node, but that only has the path as written in the Markdown file.
	- → It makes sense that this doesn't have the information.
- To access the absolute path of the markdown file itself and then figure out the image path myself. 
	- → The VFile only has the cwd which is obviously not the location of the markdown file

<details>
<summary>The information I get for an image:</summary>

```
{
  type: 'image',
  title: null,
  url: 'placeholder.jpeg',
  alt: 'Image alt text placeholder',
  position: {
    start: { line: 7, column: 1, offset: 841 },
    end: { line: 7, column: 48, offset: 888 }
  }
}
```

</details>

<details>
<summary>The the file object I get through transformer:</summary>

```
VFile {
  data: {},
  messages: [],
  history: [],
  cwd: '/Users/stephan/Documents/Projekte-kommerziell/Eileen-Website/repository',
  value: '\nLorem ipsum dolor sit amet.'
}
```

</details>

#### (B) Context:
I am using Remark through [Astro](https://astro.build/) so the issue might also arise from there.

#### (C) In [Issue 7](https://github.com/remarkjs/remark-embed-images/issues/7) of [remark-embed-images](https://github.com/remarkjs/remark-embed-images) I also read (but didn't really understand it):
> You can pass a vfile. You can also pass any argument that vfile supports (such as an object with a path).
> Without such information, plugins that use the file system have no clue where things are!

#### (D) Gatsby Plugin:
I also found [gatsby-remark-images](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-remark-images) which basically does what I need but the code seems to be tailored to Gatsby.

---

### Question:
Is something like this possible?
And if so how would I approach this?

---

**Thanks already in advance 🙏** 
(I really only post when I am completely out of ideas)

**Comment by wooorm** [comment, marked as answer]
Hi there! The answer is: it depends. First of all. This is not what remark is for. remark is about markdown. This is about HTML. You want rehype. More on the difference is available in several docs, including [this one in `remark-rehype`](https://github.com/remarkjs/remark-rehype#what-is-this).

> I have no reference to the actual image file.

Well, where *are* the images?

> I am using Remark through Astro so the issue might also arise from there.

Perhaps!

> → The VFile only has the cwd which is obviously not the location of the markdown file

It should have a `path` (which is a getter, so you don’t easily see it in the console when printing `file`, what do you see if you print `file.path`). If it doesn’t, then that’s a bug in Astro.

**Reply by stephanbogner** [reply]
Thanks a lot for the help and the super fast response ❤️ 

> Hi there! The answer is: it depends. First of all. This is not what remark is for. remark is about markdown. This is about HTML. You want rehype. More on the difference is available in several docs, including [this one in remark-rehype](https://github.com/remarkjs/remark-rehype#what-is-this).

Good call.
In my case with Astro it probably still makes sense to stick with remark. Because Astro allows to have [components inside Markdown](https://docs.astro.build/en/guides/markdown-content/#using-components-in-markdown) and also allows Markdown files [to use layouts](https://docs.astro.build/en/core-concepts/layouts/#markdown-layouts).
So basically I only want to alter the parts about images ... even though it might not be best-practice as its html inside markdown.

> Well, where are the images?

In my case they are right next to the .md-files:
E.g. `![Some alt](my-image.png)`.
In the future potentially also in a sub-folder, e.g. `![Some alt](media/my-image.png)`.

> It should have a path (which is a getter, so you don’t easily see it in the console when printing file, what do you see if you print file.path). If it doesn’t, then that’s a bug in Astro.

That's really helpful because it returns undefined. I am gonna check with the Astro-team and take it from there.

If I make some progress I am gonna post it here 🤞

**Reply by wooorm** [reply]
> In my case with Astro it probably still makes sense to stick with remark.

remark and rehype are used *in combination* in Astro. It’s not one or the other for the source language. It’s both when turning markdown into HTML. rehype is the ecosystem for things that deal with HTML, even if the original language is MD(X).

> Because Astro allows to have components inside Markdown and also allows Markdown files to use layouts.

Right, I believe I understand, as they reuse the MDX extensions to markdown I made, for their format. Those components can be access in the markdown AST and in the HTML AST. But, I believe your question is not about components, but about elements?

> I am gonna check with the Astro-team and take it from there.

:+1:

The problem seems to be here: https://github.com/withastro/astro/blob/de123b28b3ff398b800cb598f20326ca85a0fb60/packages/markdown/remark/src/index.ts#L90.

**Reply by stephanbogner** [reply]
> remark and rehype are used in combination in Astro. It’s not one or the other for the source language. It’s both when turning markdown into HTML. rehype is the ecosystem for things that deal with HTML, even if the original language is MD(X).

I see. Interesting to start to understand how it all works together. BTW: Apart from the Astro bug it seems really easy to create plugins for remark, so 👍 for your work.

> Right, I believe I understand, as they reuse the MDX extensions to markdown I made, for their format. Those components can be access in the markdown AST and in the HTML AST. But, I believe your question is not about components, but about elements?

Interesting and yes, I just want to alter the normal markdown images à la `![]()`.

> The problem seems to be here: https://github.com/withastro/astro/blob/de123b28b3ff398b800cb598f20326ca85a0fb60/packages/markdown/remark/src/index.ts#L90.

Thanks for the investigation! I [filed a bug report](https://github.com/withastro/astro/issues/3264). Let's see 🙂

**Reply by stephanbogner** [reply]
Conclusion:
It was an issue with Astro, which [was fixed now](https://github.com/withastro/astro/issues/3264) and now `file.path` works as intended.

Thanks @wooorm ❤️

**Reply by wooorm** [reply]
sweet!

**Comment by kaiwen-wang** [comment]
In the function header for the remark plugin

`    return function (tree, file) {`

Use `file.filename`
